### PR TITLE
Manage payment methods

### PR DESCRIPTION
**Pull request by huwshimi**

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- the only one that can easily be QAed is getting the payment methods, run this in your browser console: `app.payment.getPaymentMethods('YOUR USERNAME HERE', (error, data) => {console.log(error, data)});`

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7113/

File: jujugui/static/gui/src/app/jujulib/payment.js [hunk 1]
```diff
@@ -138,7 +138,6 @@ var module = module;
             - phones {Array} a list of phone number strings
         - allowEmail {Boolean} Whether the user allows emails
         - token {String|Null} A Stripe token
-        - paymentMethodName {String|Null} The name of the payment method
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/payment.js]: It does not seem this was removed.

File: jujugui/static/gui/src/app/jujulib/payment.js [hunk 2]
```diff
@@ -193,33 +192,174 @@ var module = module;
...
 
     /**
+      Get a list of the user's payment methods.
+
+      @public getPaymentMethods
+      @param username {String} The user's username.
+      @param callback {Function} A callback to handle errors or accept the
+        data from the request. Must accept an error message or null as its
+        first parameter and the payment methods as its second. The payment
+        methods returns an array of payment method objects containing:
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/payment.js]: s/returns/is/

File: jujugui/static/gui/src/app/jujulib/payment.js [hunk 3]
```diff
@@ -193,33 +192,174 @@ var module = module;
...
+            - postcode {String|Null} The address post code
+            - countryCode {String|Null} The address country code
+          - brand {String} The card brand name
+          - last4 {String} The last four digits of the card number
+          - month {Int} The card expiry month
+          - name {String} The user provided identifier of the card
+          - cardHolder {String} The name of the card owner
+          - valid {Boolean} Whether the card is valid e.g. the card has not
+            expired
+          - Year {Int} The card expiry year
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/payment.js]: s/Year/year/ and maybe list this before month?

File: jujugui/static/gui/src/app/jujulib/payment.js [hunk 4]
```diff
@@ -193,33 +192,174 @@ var module = module;
...
+    /**
+      Remove a new payment method.
+
+      @public removePaymentMethod
+      @param name {String} The user's username.
+      @param id {String} The payment method id.
+      @param token {String} A Stripe token.
+      @param callback {Function} A callback to handle errors or accept the
+        data from the request. Must accept an error message or null as its
+        first parameter and the response as the second parameter.
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/payment.js]: What's in the response? Do we care about that? Could the callback just accept an error or null? If not, let's please describe and parse the response accordingly.

File: jujugui/static/gui/src/app/jujulib/test-payment.js [hunk 5]
```diff
@@ -363,4 +370,221 @@ describe('jujulib payment service', function() {
...
+            'http://1.2.3.4/' +
+            window.jujulib.paymentAPIVersion +
+            '/u/spinach/payment-methods');
+          const xhr = makeXHRRequest({
+            'payment-methods': [{
+              address: {
+                name: 'Home',
+                line1: '1 Maple St',
+                line2: null,
+                county: 'Bunnyhug',
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: Who doesn't love bunnyhugs?

File: jujugui/static/gui/src/app/jujulib/test-payment.js [hunk 6]
```diff
@@ -363,4 +370,221 @@ describe('jujulib payment service', function() {
...
+      payment.getPaymentMethods('spinach', function(error, response) {
+        assert.strictEqual(error, null);
+        assert.deepEqual(response, [{
+          address: {
+            name: 'Home',
+            line1: '1 Maple St',
+            line2: null,
+            state: 'Bunnyhug',
+            city: 'Sasquatch',
+            postcode: '90210',
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: IIRC from my teenage, this zip code does not belong to sasquatch.

File: jujugui/static/gui/src/app/jujulib/test-payment.js [hunk 7]
```diff
@@ -363,4 +370,221 @@ describe('jujulib payment service', function() {
...
+      const bakery = {
+        sendGetRequest: function(path, success, failure) {
+          const xhr = makeXHRRequest({Message: 'Uh oh!'});
+          failure(xhr);
+        }
+      };
+      const payment = new window.jujulib.payment('http://1.2.3.4/', bakery);
+      payment.getPaymentMethods('spinach', function(error, user) {
+        assert.equal(error, 'Uh oh!');
+        assert.isNull(user);
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: `assert.strictEqual(user, null);` ?

File: jujugui/static/gui/src/app/jujulib/test-payment.js [hunk 8]
```diff
@@ -363,4 +370,221 @@ describe('jujulib payment service', function() {
...
+      const bakery = {
+        sendPutRequest: function(path, params, success, failure) {
+          const xhr = makeXHRRequest({Message: 'Uh oh!'});
+          failure(xhr);
+        }
+      };
+      const payment = new window.jujulib.payment('http://1.2.3.4/', bakery);
+      payment.createPaymentMethod('spinach', 'token123', (error, response) => {
+        assert.equal(error, 'Uh oh!');
+        assert.isNull(response);
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: Ditto.

File: jujugui/static/gui/src/app/jujulib/test-payment.js [hunk 9]
```diff
@@ -363,4 +370,221 @@ describe('jujulib payment service', function() {
...
+          const xhr = makeXHRRequest('success');
+          success(xhr);
+        }
+      };
+      const payment = new window.jujulib.payment('http://1.2.3.4/', bakery);
+      payment.removePaymentMethod(
+        'spinach', 'payment-method-created-2017-3-11-13-49-42-186', 'token123',
+        (error, response) => {
+          assert.isNull(error);
+          assert.deepEqual(response, 'success');
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: I still think this is a bit redundant considering we already have an error.

File: jujugui/static/gui/src/app/jujulib/payment.js [hunk 10]
```diff
@@ -193,33 +192,174 @@ var module = module;
...
+      };
+      const url = `${this.url}/u/${username}/payment-methods`;
+      return jujulib._makeRequest(this.bakery, url, 'GET', null, handler);
+    },
+
+    /**
+      Create a new payment method.
+
+      @public createPaymentMethod
+      @param name {String} The user's username.
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/jujulib/payment.js]: should be username, same with some of the other doc

File: jujugui/static/gui/src/app/jujulib/payment.js [hunk 11]
```diff
@@ -193,33 +192,174 @@ var module = module;
...
+      };
+      return jujulib._makeRequest(this.bakery, url, 'DELETE', payload, handler);
+    },
+
+    /**
+      Generate an ID of this payment method. The ID only needs to be unique per
+      user so using the full timestamp should be enough.
+
+      @public _generatePaymentMethodName
+      @param paymentMethods {Array} Payment method reponses from the API.
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/jujulib/payment.js]: Param doesn't exist

File: jujugui/static/gui/src/app/jujulib/payment.js [hunk 12]
```diff
@@ -193,33 +192,174 @@ var module = module;
...
+      ];
+      return `payment-method-created-${parts.join('-')}`;
+    },
+
+    /**
+      Reformat payment method objects for easier use with JavaScript.
+
+      @public _parsePaymentMethods
+      @param paymentMethods {Array} Payment method reponses from the API.
+      @returns {Array} A list of parsed payment method object.
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/jujulib/payment.js]: Second param `parseCareholder` not declared

File: jujugui/static/gui/src/app/jujulib/test-payment.js [hunk 13]
```diff
@@ -363,4 +370,221 @@ describe('jujulib payment service', function() {
...
+        }
+      };
+      const payment = new window.jujulib.payment('http://1.2.3.4/', bakery);
+      payment.createPaymentMethod('spinach', 'token123', (error, response) => {
+        assert.strictEqual(error, null);
+        assert.deepEqual(response, {
+          address: {
+            name: 'Home',
+            line1: '1 Maple St',
+            line2: null,
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: Maybe mix it up a bit for the test and add some `line2`'s for some of them? Not sure if it'll catch anything/ make a difference.

File: jujugui/static/gui/src/app/jujulib/test-payment.js (hunk 5 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: +1

File: jujugui/static/gui/src/app/jujulib/test-payment.js (hunk 6 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: But a lot of fabulous people live in that zip code @frankban. @huwshimi just adding some razzle dazzle.

File: jujugui/static/gui/src/app/jujulib/test-payment.js (hunk 7 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: isNull is a strictEqual null isn't it?

Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/test-payment.js]: The latter used to provide more information, and I didn't receive any news that it changed. This involves at least isUndefined, IsNull, isTrue and isFalse.

File: jujugui/static/gui/src/app/jujulib/payment.js (hunk 1 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/jujulib/payment.js]: Ah yeah, we no longer need to pass it in now as we generate it automatically.

File: jujugui/static/gui/src/app/jujulib/payment.js (hunk 4 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/jujulib/payment.js]: That's a great question and the answer is I haven't been able to test it yet so I don't know! I was planning to update that once I have the implementation working (the API doc doesn't have any details).

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7126/

**Comment by huwshimi** [comment]
Thanks @frankban @Lukewh :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

File: jujugui/static/gui/src/app/jujulib/payment.js (hunk 4 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/jujulib/payment.js]: You were right, I've updated this in my current branch to remove the unused param.

**Pull request merged**